projects/cds/src/profiletag/services/profile-tag-push-events.service.ts

```diff
@@ -0,0 +1,150 @@
+import { Injectable } from '@angular/core';
+import {
+  ActiveCartService,
+  EventService,
+  PersonalizationContextService,
+} from '@spartacus/core';
+import {
+  CartPageVisited,
+  CategoryPageVisited,
+  HomePageVisited,
+  KeywordSearchPageVisited,
+  OrderConfirmationPageVisited,
+  PageVisited,
+  ProductDetailsPageVisited,
+} from 'projects/core/src/routing/event/routing.events';
+import { merge, Observable, of } from 'rxjs';
+import { map, skipWhile, withLatestFrom } from 'rxjs/operators';
+import {
+  CartChangedPushEvent,
+  CartViewPushEvent,
+  CategoryViewPushEvent,
+  HomePageViewPushEvent,
+  KeywordSearchPushEvent,
+  OrderConfirmationPushEvent,
+  PageViewPushEvent,
+  ProductViewPushEvent,
+  ProfileTagPushEvent,
+} from '../model/profile-tag.model';
+
+@Injectable({
+  providedIn: 'root',
+})
+export class ProfileTagPushEventsService {
+  protected pushEvents$: Observable<ProfileTagPushEvent> = merge(
+    this.categoryPageVisited(),
+    this.productDetailsPageView(),
+    this.searchResultsChanged(),
+    this.homePageVisitedEvent(),
+    this.cartPageVisitedEvent(),
+    this.pageVisitedEvent(),
+    this.orderConfirmationVisited(),
+    this.cartChanged()
+  );
+  constructor(
+    protected activeCartService: ActiveCartService,
+    protected eventService: EventService,
+    protected personalizationContextService: PersonalizationContextService
+  ) {}
+
+  getPushEvents(): Observable<ProfileTagPushEvent> {
+    return this.pushEvents$.pipe(
+      withLatestFrom(
+        merge(
+          of({ segments: undefined, actions: undefined }),
+          this.personalizationContextService.getPersonalizationContext()
+        )
+      ),
+      map(([item, personalizationContext]) => {
+        item.data = item.data ? item.data : {};
+        item.data.segments = personalizationContext.segments;
+        item.data.actions = personalizationContext.actions;
+        return item;
+      })
+    );
+  }
+
+  addPushEvent(event: Observable<ProfileTagPushEvent>) {
+    this.pushEvents$ = merge(this.pushEvents$, event);
+  }
+
+  /**
+   * Listens to the changes to the cart and pushes the event for profiletag to pick it up further.
+   */
+  cartChanged(): Observable<CartChangedPushEvent> {
```

Comment:
Is this supposed to be `public`? Maybe it should also be `protected`?

Reply:
We made it protected and adapted the failing tests accordingly.